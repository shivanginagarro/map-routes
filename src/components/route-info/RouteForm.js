import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

import { MAP_SETTING } from "../map/map-config";
import { mapInitOnLoad } from '../map/map-api-init-service';
var GoogleMapsLoader = require('google-maps');

let _this;
class RouteForm extends React.PureComponent {

    startPointPlaces;
    dropPointPlaces;

    constructor(props) {
        super(props);
        _this = this;        
        _this.state = {
            startingPoint: "",
            droppingPoint: "",
            message: ""
        }
    }

    componentWillMount() {
        _this.setUpApiKeyInMap();
    }

    setUpApiKeyInMap = async () => {
        // NOTE TO SELF :probably use destructuring if possible
        GoogleMapsLoader.KEY = MAP_SETTING.KEY;
        GoogleMapsLoader.LIBRARIES = MAP_SETTING.LIBRARIES;
        let response = await mapInitOnLoad(GoogleMapsLoader);
        _this.map = response;
        _this.getPlacesData(response);
    }


    getPlacesData = (googleObj) => {
        let originPlace = new googleObj.maps.places.Autocomplete(_this.startPoint)
        let destinationPlace = new googleObj.maps.places.Autocomplete(_this.dropPoint)
        googleObj.maps.event.addListener(originPlace, 'place_changed', () =>{
            _this.startPointPlaces = originPlace.getPlace();
         })
         googleObj.maps.event.addListener(destinationPlace, 'place_changed', () =>{
            _this.dropPointPlaces = destinationPlace.getPlace();
         })
    }


    getRouteInMap = (event) =>{
        event.preventDefault();
        this.props.getRoutes(_this.startPointPlaces,_this.dropPointPlaces);     
    }

    render() {
        return (
            <form>
                <label> Starting Point:  <input type="text" name="starting" ref={el => _this.startPoint = el} placeholder="Pick Up From" /> </label>
                <label> Dropping Point:  <input type="text" name="dropping" ref={el => _this.dropPoint = el} placeholder="Drop To" /> </label>
                {_this.state.message && _this.state.message.length > 0 ? <div>{_this.state.message}</div> : null}
                <input type="submit" value="Submit" onClick={_this.getRouteInMap} /> <button>Reset</button>
            </form>
        )
    }
}

RouteForm.propTypes = {
    googleObject: PropTypes.object
}
export default RouteForm;
