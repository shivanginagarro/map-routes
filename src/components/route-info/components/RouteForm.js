import React from 'react';
import PropTypes from 'prop-types';
import '../route-style.css';
import RouteInfo from "./RouteInfo";
import { mapInitOnLoad } from "../../../common/services/index";
import { MAP_SETTING } from "../../../common/configurations/constants/index";

class RouteForm extends React.PureComponent {
    //to save selected place object for origin 
    startPointPlaces;
    //to save selected place object for destination     
    dropPointPlaces;
    //to store google object locally
    google;

    constructor(props) {
        super(props);
        this.state = {
            startingPoint: "",
            droppingPoint: "",
            error: null,
            total_distance: null,
            total_time: null,
            labelSubmit:"Submit"
        }
    }

    //call to get google map object
    async setUpApiKeyInMap(){
        let response = await mapInitOnLoad();
        this.google = response;
        this.getPlacesData(response);        
    }

    componentDidMount() {
          this.setUpApiKeyInMap();
    }

    /**
     * @description
     * to get origin and destination dropdown options for places in input fields respectively.
     * Autocomplete method of google api is used to get it   
     */
    getPlacesData = (googleObj) => {
        let originPlace = new googleObj.maps.places.Autocomplete(this.startPoint)
        let destinationPlace = new googleObj.maps.places.Autocomplete(this.dropPoint)
        googleObj.maps.event.addListener(originPlace, 'place_changed', () => {
            this.startPointPlaces = originPlace.getPlace();
        })
        googleObj.maps.event.addListener(destinationPlace, 'place_changed', () => {
            this.dropPointPlaces = destinationPlace.getPlace();
        })
        }

    /**
     * @description
     * to get latitude and longitude of seleted origin and destination places and
     *  then make a request to backend to get the coordinates of routes/path  
     */
    getRouteInMap = (event) => {
        event.preventDefault();
        //resetting states but not working need to check.
        this.props.reset();
        let  originLocation =  this.startPointPlaces && this.startPointPlaces.geometry.location;
        let  destinationLocation = this.dropPointPlaces && this.dropPointPlaces.geometry.location; 
        let originCoordinates = [
            originLocation.lat().toString(),originLocation.lng().toString()
        ]
        let destinationCorodinates = [
            destinationLocation.lat().toString(),destinationLocation.lng().toString()
        ]
        this.props.getRoutes({ "origin": originCoordinates, "destination": destinationCorodinates })
    }
    /**
     * @description
     * reset the entire form :reset the inputdivs and reset the information div of our form on click of reset button
     */
    resetForm = () => {
        this.startPoint.value = null;
        this.dropPoint.value = null;
        this.props.reset();
    }

    //updates states on getting any changes in props value
    static getDerivedStateFromProps(props,state) {
        const { informationDetail } = props;
        const { total_time, total_distance, error } = informationDetail;
        if (informationDetail) {
            return {
                total_distance,
                total_time,
                error,
                labelSubmit: total_distance || error ?"Re-Submit":"Submit"
            }
        }
        return null;
    }


    render() {
        return (
            <div className="directions-form">
                <label className="form-label"> <span>Starting Point :</span>  <input type="text" name="starting" ref={el => this.startPoint = el} placeholder="Pick Up From" /> </label>
                <label className="form-label"> <span>Dropping Point :</span>  <input type="text" name="dropping" ref={el => this.dropPoint = el} placeholder="Drop To" /> </label>
                <RouteInfo informationDetail={this.props.informationDetail}/>
                <div className="btn-container">
                    <button className="form-btn" onClick={this.getRouteInMap}>{this.state.labelSubmit}</button> <button onClick={this.resetForm}>Reset</button>
                </div>
            </div>
        )
    }
}

RouteForm.protoType = {
    informationDetail : PropTypes.object.isRequired,
    getRoutes: PropTypes.func.isRequired,
    reset:PropTypes.func.isRequired
}

export default RouteForm;
