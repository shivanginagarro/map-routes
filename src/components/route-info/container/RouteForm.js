import React from 'react';
import PropTypes from 'prop-types';
import '../route-style.css';
import RouteInfo from "../component/RouteInfo";
import { mapInitOnLoad } from "../../../common/services/index";
import { requiredValidation, samePlaceValidation } from "../constant/constant";

class RouteForm extends React.PureComponent {
    //to save selected place object for origin 
    startPointPlaces;
    //to save selected place object for destination     
    dropPointPlaces;
    //to store google object locally
    google;

    constructor(props) {
        super(props);
        this.state = {
            error: null,
            total_distance: null,
            total_time: null,
            labelSubmit: "Submit",
            origin: '',
            destination: ''
        }
        this.originChangeHanlder = this.originChangeHanlder.bind(this);
        this.destinationChangeHanlder = this.destinationChangeHanlder.bind(this)
    }

    //call to get google map object
    async setUpApiKeyInMap() {
        let response = await mapInitOnLoad();
        this.google = response;
        this.getPlacesData(response);
    }

    componentDidMount() {
        this.setUpApiKeyInMap();
    }

    /**
     * @description
     * to get origin and destination dropdown options for places in input fields respectively.
     * Autocomplete method of google api is used to get it   
     */
    getPlacesData = (googleObj) => {
        this.startPoint = document.getElementById("origin");
        this.dropPoint = document.getElementById("destination");
        let originPlace = new googleObj.maps.places.Autocomplete(this.startPoint)
        let destinationPlace = new googleObj.maps.places.Autocomplete(this.dropPoint)
        googleObj.maps.event.addListener(originPlace, 'place_changed', () => {
            this.startPointPlaces = originPlace.getPlace();
            this.setState({ origin: this.startPointPlaces.formatted_address });
        })
        googleObj.maps.event.addListener(destinationPlace, 'place_changed', () => {
            this.dropPointPlaces = destinationPlace.getPlace();
            this.setState({ destination: this.dropPointPlaces.formatted_address });
        })
    }

    /**
     * @description
     * to get latitude and longitude of seleted origin and destination places and
     *  then make a request to backend to get the coordinates of routes/path  
     */
    getRouteInMap = (event) => {
        event.preventDefault();
        const { origin, destination } = this.state;
        //validation for valid input
        if (!origin || !destination) {
            this.props.validationHandling({
                error: requiredValidation
            });
            return;
        }
        if (origin === destination) {
            this.props.validationHandling({
                error: samePlaceValidation
            });
            return;
        }
        //resetting states but not working need to check.
        this.props.reset();
        let originLocation = this.startPointPlaces && this.startPointPlaces.geometry.location;
        let destinationLocation = this.dropPointPlaces && this.dropPointPlaces.geometry.location;
        let originCoordinates = [
            originLocation.lat().toString(), originLocation.lng().toString()
        ]
        let destinationCorodinates = [
            destinationLocation.lat().toString(), destinationLocation.lng().toString()
        ]
        this.props.getRoutes({ "origin": originCoordinates, "destination": destinationCorodinates })
    }
    /**
     * @description
     * reset the entire form :reset the inputdivs and reset the information div of our form on click of reset button
     */
    resetForm = () => {
        this.setState({
            origin:'',
            destination:''
        })
        this.props.reset();
    }

    //updates states on getting any changes in props value
    static getDerivedStateFromProps(props, state) {
        const { informationDetail } = props;
        const { total_time, total_distance, error } = informationDetail;
        if (informationDetail) {
            return {
                total_distance,
                total_time,
                error,
                labelSubmit: total_distance || error ? "Re-Submit" : "Submit"
            }
        }
        return null;
    }

    originChangeHanlder(event) {
        this.setState({ origin: event.target.value });
    }

    destinationChangeHanlder(event) {
        this.setState({ destination: event.target.value });
    }

    render() {
        return (
            <div className="directions-form">
                <label className="form-label">
                    <span>Starting Point :</span>
                    {this.state.origin.length > 0 && <span className={"float-right"} onClick={() => this.setState({ origin: "" })}><img height={"20px"} width={"20px"} alt="Remove" src={require("../../../images/remove.png")} /></span>}
                    <input type="text" value={this.state.origin} onChange={this.originChangeHanlder} name="starting" id={"origin"} placeholder="Pick Up From" />
                </label>
                <label className="form-label">
                    <span>Dropping Point :</span>
                    {this.state.destination.length > 0 && <span className={"float-right"} onClick={() => this.setState({ destination: "" })}><img height={"20px"} width={"20px"} alt="Remove" src={require("../../../images/remove.png")} /></span>}
                    <input type="text" value={this.state.destination} onChange={this.destinationChangeHanlder} name="dropping" id={"destination"} placeholder="Drop To" />
                </label>
                <RouteInfo informationDetail={this.props.informationDetail} />
                <div className="btn-container">
                    <button className="form-btn" onClick={this.getRouteInMap}>{this.state.labelSubmit}</button> <button onClick={this.resetForm}>Reset</button>
                </div>
            </div>
        )
    }
}

RouteForm.protoType = {
    informationDetail: PropTypes.object.isRequired,
    getRoutes: PropTypes.func.isRequired,
    reset: PropTypes.func.isRequired
}

export default RouteForm;
