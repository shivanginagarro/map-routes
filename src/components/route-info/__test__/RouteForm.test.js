import React from "react";
import RouteForm from "../container/RouteForm";
import renderer from "react-test-renderer";

describe("Component:RouteForm",() => {
   it("Testing the Route Form Component",() => {
    let informationDetail = {
        error: '',
        total_distance: '',
        total_time: ''
    }
    let getRoutes = () => {
        console.log('inside routes testing');
    }
    let reset = () => {
        console.log("inside reset testing infor");
    }
    let validationHandling = () => {
        console.log("inside validationHandling");
    }
    const tree = renderer.create(<RouteForm informationDetail={informationDetail} getRoutes={getRoutes} reset={reset} validationHandling={validationHandling}/>).toJSON();
    expect(tree).toMatchSnapshot();
   });
})