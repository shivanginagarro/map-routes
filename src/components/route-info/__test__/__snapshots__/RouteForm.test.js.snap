// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:RouteForm Testing the Route Form Component 1`] = `
<div
  className="directions-form"
>
  <label
    className="form-label"
  >
    <span>
      Starting Point :
    </span>
    <input
      id="origin"
      name="starting"
      onChange={[Function]}
      placeholder="Pick Up From"
      type="text"
      value=""
    />
  </label>
  <label
    className="form-label"
  >
    <span>
      Dropping Point :
    </span>
    <input
      id="destination"
      name="dropping"
      onChange={[Function]}
      placeholder="Drop To"
      type="text"
      value=""
    />
  </label>
  <div
    className="btn-container"
  >
    <button
      className="form-btn"
      onClick={[Function]}
    >
      Submit
    </button>
     
    <button
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
</div>
`;
